[WIP] [gen] Update baseline configuration for `rmw`

ShaleXIONG opened the pull request:
Add several missing `rmw` relaxations since such relaxation can match memory access such as `M`, `R`, and `W`.

this pull requisition contains #1569 too.

chacon01 reviewed (changes requested):
Thanks @ShaleXIONG! A couple of comments.

ShaleXIONG commented on gen/libdir/forbidden.conf:
```diff
@@ -71,80 +71,157 @@
+# `[DpAddrCseldR rmw PosWW]` can be combined by `diy` as `[DpAddrCseldR rmw]` and `PosWW`
+-safe [DpAddrCseldR PodRR Amo.Swp] [DpAddrCseldR PosRR Amo.Swp] [DpAddrCseldW PodWR Amo.Swp] [DpAddrCseldW PosWR Amo.Swp] [DpAddrCselsR PodRR Amo.Swp] [DpAddrCselsW PodWR Amo.Swp]
+-safe [DpAddrCseldR PodRR Amo.Cas] [DpAddrCseldR PosRR Amo.Cas] [DpAddrCseldW PodWR Amo.Cas] [DpAddrCseldW PosWR Amo.Cas] [DpAddrCselsR PodRR Amo.Cas] [DpAddrCselsW PodWR Amo.Cas]
+-safe [DpAddrCseldR PodRR LxSx] [DpAddrCseldR PosRR LxSx] [DpAddrCseldW PodWR LxSx] [DpAddrCseldW PosWR LxSx] [DpAddrCselsR PodRR LxSx] [DpAddrCselsW PodWR LxSx]
```
I will manually add all the components.

Commit:
[gen] Add `ExpObs`,`InpObs` and `ExpInpObs` wildcard for ext- int- and both-type- communication.

Commit:
[gen] Strengthen forbidden.conf for less locally-order-before litmus tests.

For `...;po` and `...;po;[M]` in cat, we append the external
communication edges `Obs` in the translation. Similarly for `po;...` and
`[M];po;...`, we prefix `Obs`.

Commit:
[gen] Add AArch64 MemTag generator config

Add gen/libdir/memtag.conf for generating forbidden tests for memory
tags.

Commit:
[gen] Add more none fault checks.

Add more consecutive none fault check. This rule out undesired situations
where a memory access does not fault but the consecutive one faults.

Commit:
[gen] Move label reset to the begining of cycle building.

Commit:
[gen] add back Amo.StAdd and Amo.LdAdd in baseline configuration file.

Add back `Amo.StAdd` and `Amo.LdAdd` as the representation for all other
atomic operation. We also remove `[Q Amo.* ***]` as `Q` cannot apply to
`Amo.*`

Commit:
[gen] Add more rmw to match memory access in `forbidden.conf`.

chacon01 commented on gen/libdir/forbidden.conf:
```diff
@@ -205,72 +272,77 @@
 
 # let bob = [Exp & M | Imp & Tag & R]; po; [L]
--safe [Pod*W L]
+-safe [ExpObs Pod*W L]
```
Similar to the acquire case above, this won't compose with `Amo` or `LxSx`. In addition to these, we now need
* `[ExpObs Amo|LxSx PodWW L]`

chacon01 commented on gen/libdir/forbidden.conf:
```diff
@@ -205,72 +272,77 @@
+-safe [A Amo.Cas L PodW* ExpObs] [A Amo.Cas L PosWR ExpObs]
+-safe [A Amo.StAdd L PodW* ExpObs] [A Amo.StAdd L PosWR ExpObs]
+-safe [A Amo.LdAdd L PodW* ExpObs] [A Amo.LdAdd L PosWR ExpObs]
+-safe [A LxSx L PodW* ExpObs]    [A LxSx L PosWR ExpObs]
```
`[range([A];amo;[L])]; po; [Exp & M | Imp & Tag & R | MMU & FAULT]` does not apply for `LxSx`, so this relaxation can be removed. While this will not produce allowed tests because we would need it to compose with a free `Pod*R` (that doesn't exist), it is inconsistent with the justifying comment.

For the `Amo` variants (and `LxSx` too), the generator will only compose these with the other relaxations which are properly ordered into the acquire read effect, so all the forbidden combinations will also be forbidden without the `L`.

chacon01 commented on gen/libdir/forbidden.conf:
```diff
@@ -205,72 +272,77 @@
+-safe [ExpObs Po**R A Amo.Swp L Po*W* ExpObs]
+-safe [ExpObs Po**R A Amo.Cas L Po*W* ExpObs]
+-safe [ExpObs Po**R A Amo.StAdd L Po*W* ExpObs]
+-safe [ExpObs Po**R A Amo.LdAdd L Po*W* ExpObs]
```
The `ExpObs` here prevents these from composing with `Amo` or `LxSx` so they need to be specified individually.
* `[ExpObs Amo|LxSx Po*WR Amo L Po*W* ExpObs]`
* `[ExpObs Po**R Amo L Po*WR Amo|LxSx ExpObs]`
* `[ExpObs Amo|LxSx Po*WR Amo L Po*WR Amo|LxSx ExpObs]`

chacon01 reviewed (changes requested):
`Amo` and `LxSx` become tedious to enumerate the more the configuration file is restricted to prevent additional ordering. I see two issues:
* Several cases where an existing effect could be an `Amo` or `LxSx` without changing the meaning
* We can't create a general `Rmw` or `Amo` relaxation because it will include e.g. `Amo.LdClr` which has the value collision problem.

Perhaps the generator could support "user-defined" relaxations to help here e.g.
```
-define LdRmw LxSx|Amo.Swp|Amo.Cas|Amo.LdAdd
-define Rmw LdRmw|Amo.StAdd
...
-safe [A LdRmw PodW* ExpObs]
-safe [ExpObs PodWR Rmw L]
```
which would reduce the duplication everywhere else.

Additionally, perhaps an operation to make a relaxation "optional" e.g.
```
-safe [A Rmw? Pod** Rmw? ExpObs]
```
which would be equivalent to
```
-safe [A Pod** ExpObs]
-safe [A Rmw Pod** ExpObs]
-safe [A Pod** Rmw ExpObs]
-safe [A Rmw Pod** Rmw ExpObs]
```

ShaleXIONG committed:
[gen] Add predicate such as `@after(..)` syntax to relax parser

Extend the relax AST with a predicate node and teach the lexer/parser to
recognise predicate applications such as `@before(...)` and `@after(...)`.

Add `map_predicate` support to the AST so predicate names can be
converted before expansion. Parse `before` and `after` into edge predicates,
then attach the parsed predicate to every edge produced by the decorated
relax expression.

ShaleXIONG committed:
[gen] Strenth the baseline configuration by before and after predicate.

ShaleXIONG committed:
[gen] Filter invalid diy7 predicate relaxations

Move predicate-specific relaxation filtering into Alt.Make, where diy7
predicate edges are represented.

Reject predicate-only relaxations and invalid before/after predicate
placement after expanding the plain relaxation through the existing
Relax.remove_invalid_relaxes path. Expose the predicate-aware filtering
helper for parser tests.

ShaleXIONG committed:
[gen] Update CI due to the new version of baseline.

ShaleXIONG committed:
[gen] refactor DumpAll test preparation

Split DumpAll's check/dump path into explicit and clear stages: atom variation,
edge resolution, normalisation, duplicate detection, test construction, and
final dumping.  This makes the ordering clearer and avoids rebuilding the same
test separately for each scope/name variant. It also massively simply error
handling.

minor:
- Change `mk_info` from a function over the normalised edge list into prepared
  metadata passed with the candidate.
- Make `Edge.varatom` return the list of atomic variations directly, leaving
  edge resolution to the DumpAll pipeline. This keeps varatom focused on atom
  enumeration and makes resolve_edges placement explicit.

ShaleXIONG committed:
[gen] filter invalid cycles in `diy7` searching algorithm before DumpAll

Add earlier candidate checks in alt.ml before calling into DumpAll/test
construction
- It rejects candidates that already exceed the configured process or
  per-procedure instruction limit
- It checks both (1) the immediate adjacent boundary and (2) the effective
  non-pseudo memory-access boundary when concatenating relaxation.
- Rework on rejecting cycle or cycle segment.
  Replace the old reject substring/span helpers with direct cycle/list
  matching helpers used by both partial generation and last-minute checking.
  For the final check, we apply a simpler and more readable algorithm
  by checking if the `cycle @ cycle` contains any reject edge lists, which
  is much more efficient and readable compated to previous `cut` based.

ShaleXIONG committed:
[gen] precompute adjacency checks during cycle search

Add an Adjacent helper for the diy7 cycle search that assigns local ids to
relaxation chunks and precomputes their adjacency compatibility.

This lets the recursive search use table lookups for repeated predecessor
checks instead of recomputing the filter and boundary predicate merge logic
on every call.

ShaleXIONG committed:
[gen] Add store-only variant

Add the `StoreOnly` generator variant and propagate it to herd metadata as
`store-only`.

ShaleXIONG committed:
[gen] Carry `store-only` fault labels to stores

In `store-only` mode, fault labels should skip loads and attach to the
first store instruction instead. This applies to ordinary load/store
sequences as well as `LxSx` sequences on `AArch64`, where the label must
be carried from the exclusive load to the following exclusive store.

ShaleXIONG committed:
[gen] restrict read value checks to communication-adjacent events

Only mark read values for final-condition checking when the event is
adjacent to a communication edge and the location already requires value
checking.

ShaleXIONG committed:
[gen] Change CSEL dependency instruction sequences

Generate AArch64 CSEL dependency variants using register self-compare
instead of hardcode comparison against zero. This makes the
selected CSEL path deterministic. It also removes the extra register
instruction.

Update the affected AArch64 generator baselines.

ShaleXIONG committed:
[gen] Use explicit equality helpers for edge comparison

Add equality helpers for code, edge components, dependencies, and RMWs.
This avoids polymorphic equality for `tedge` internals.
Use existing `compare_atom` for `atom` equality check.

ShaleXIONG committed:
[gen] Process before/after predicates during diy7 search

Teach the diy7 search to interpret leading `@before(...)` and trailing
`@after(...)` predicate edges when deciding whether two relaxation
fragments can be placed next to each other.

ShaleXIONG committed:
[gen] clarify cycle search budget helper functions.

ShaleXIONG committed:
[gen] Simplify the diy7 searching budget for `-size`.

It now always decreases by `1`. Previously, it only decreases when the
input relax contain at least non-pseudo edge. This is unnecessary.
